File: typescript/server/test/user-scoping.test.ts
/**
 * Per-user conversation scoping — the cross-user data leak (pearl th-8fe998).
 *
 * Before this, `listConversations` took no user filter, so `list_conversations`
 * returned EVERY user's conversations, and neither the resume path nor
 * `get_conversation_messages` checked ownership. Any authenticated user could
 * enumerate and open anyone else's chats.
 *
 * These tests are written from the ATTACKER's side: user A, authenticated, trying
 * to see or touch user B's conversations by every route the protocol offers —
 * listing, resuming, reading messages, posting turns, and lying about their
 * identity in the create frame. The most important assertion in the file is the
 * existence-oracle one: B's conversation and an id that never existed must produce
 * IDENTICAL payloads, or the difference itself leaks which ids are real.
 */
import { MockLlmProvider } from '@smooai/smooth-operator-core';
import { describe, expect, it } from 'vitest';

import type { AccessContext } from '../src/auth.js';
import { ANONYMOUS_ACCESS } from '../src/auth.js';
import { FrameDispatcher } from '../src/frameDispatcher.js';
import type { Frame } from '../src/protocol.js';
import { InMemorySessionStore } from '../src/sessionStore.js';

/** An authenticated principal on an auth-ENABLED server, optionally with no email claim. */
function principal(email: string | undefined): AccessContext {
    return {
        principal: { sub: 'user-sub', org: 'acme', role: 'basic', groups: [], ...(email ? { email } : {}) },
        isAnonymous: false,
        authEnabled: true,
    };
}

/** Anonymous on an auth-ENABLED server — a missing/expired/forged token lands here. */
const ANON_UNDER_AUTH: AccessContext = { ...ANONYMOUS_ACCESS, authEnabled: true };

/** A dispatcher bound to `access`, over a SHARED store, plus a sink-bound dispatch. */
function connect(store: InMemorySessionStore, access: AccessContext) {
    const dispatcher = new FrameDispatcher({ store, chatClient: new MockLlmProvider(), access });
    const sink: Frame[] = [];
    const dispatch = (frame: Record<string, unknown>) => dispatcher.dispatch(JSON.stringify(frame), (f) => sink.push(f));
    return { sink, dispatch };
}

/** Create a conversation owned by `email` with one message, via the create action. */
async function seed(store: InMemorySessionStore, email: string, text: string) {
    const { sink, dispatch } = connect(store, principal(email));
    await dispatch({ action: 'create_conversation_session', requestId: 'cs', agentId: 'agent-1' });
    const data = sink[0]!.data as { sessionId: string; conversationId: string };
    await store.appendMessage(data.conversationId, 'inbound', text);
    return data;
}

const conversationIds = (sink: Frame[]): string[] =>
    ((sink[0]!.data as { conversations: { conversationId: string }[] }).conversations ?? []).map((c) => c.conversationId);

describe('per-user conversation scoping (th-8fe998)', () => {
    describe('list_conversations', () => {
        it("returns ONLY the caller's conversations, never another user's", async () => {
            const store = new InMemorySessionStore();
            const a = await seed(store, 'a@example.com', 'alice secret');
            const b = await seed(store, 'b@example.com', 'bob secret');

            const { sink, dispatch } = connect(store, principal('a@example.com'));
            await dispatch({ action: 'list_conversations', requestId: 'lc' });

            expect(conversationIds(sink)).toEqual([a.conversationId]);
            expect(conversationIds(sink)).not.toContain(b.conversationId);
        });

        it('scopes inside the selection, so a small limit still returns the caller\'s own rows', async () => {
            // The filter-after-limit bug: 5 of B's conversations are newer than A's one,
            // so a limit of 2 applied BEFORE the owner filter yields an empty page for A.
            const store = new InMemorySessionStore();
            const a = await seed(store, 'a@example.com', 'alice');
            for (let i = 0; i < 5; i++) await seed(store, 'b@example.com', `bob ${i}`);

            const { sink, dispatch } = connect(store, principal('a@example.com'));
            await dispatch({ action: 'list_conversations', requestId: 'lc', limit: 2 });

            expect(conversationIds(sink)).toEqual([a.conversationId]);
        });

        it('fails closed to an EMPTY list when auth is on but the principal has no email', async () => {
            const store = new InMemorySessionStore();
            await seed(store, 'a@example.com', 'alice');
            await seed(store, 'b@example.com', 'bob');

            for (const access of [principal(undefined), ANON_UNDER_AUTH]) {
                const { sink, dispatch } = connect(store, access);
                await dispatch({ action: 'list_conversations', requestId: 'lc' });
                // Never a silent fall back to unscoped.
                expect(conversationIds(sink)).toEqual([]);
            }
        });

        it('stays UNSCOPED when auth is disabled (single-tenant local/dev)', async () => {
            const store = new InMemorySessionStore();
            const s1 = await store.createSession('agent-1', 'One', 'one@example.com');
            const s2 = await store.createSession('agent-1', 'Two', 'two@example.com');
            await store.appendMessage(s1.conversationId, 'inbound', 'one');
            await store.appendMessage(s2.conversationId, 'inbound', 'two');

            const { sink, dispatch } = connect(store, ANONYMOUS_ACCESS);
            await dispatch({ action: 'list_conversations', requestId: 'lc' });

            expect(conversationIds(sink).sort()).toEqual([s1.conversationId, s2.conversationId].sort());
        });
    });

    describe("reading another user's session", () => {
        it('get_conversation_messages on B\'s session is SESSION_NOT_FOUND', async () => {
            const store = new InMemorySessionStore();
            const b = await seed(store, 'b@example.com', 'bob secret');

            const { sink, dispatch } = connect(store, principal('a@example.com'));
            await dispatch({ action: 'get_conversation_messages', requestId: 'gm', sessionId: b.sessionId });

            expect(sink[0]!.type).toBe('error');
            expect((sink[0]!.error as Record<string, unknown>).code).toBe('SESSION_NOT_FOUND');
        });

        it('get_session on B\'s session is SESSION_NOT_FOUND (no agent/conversation ids leak)', async () => {
            const store = new InMemorySessionStore();
            const b = await seed(store, 'b@example.com', 'bob secret');

            const { sink, dispatch } = connect(store, principal('a@example.com'));
            await dispatch({ action: 'get_session', requestId: 'gs', sessionId: b.sessionId });

            expect(sink[0]!.type).toBe('error');
            expect(JSON.stringify(sink[0]!)).not.toContain(b.conversationId);
        });

        it('send_message into B\'s session is SESSION_NOT_FOUND (no writing to another user\'s chat)', async () => {
            const store = new InMemorySessionStore();
            const b = await seed(store, 'b@example.com', 'bob secret');

            const { sink, dispatch } = connect(store, principal('a@example.com'));
            await dispatch({ action: 'send_message', requestId: 'sm', sessionId: b.sessionId, message: 'injected' });

            expect(sink[0]!.type).toBe('error');
            expect((sink[0]!.error as Record<string, unknown>).code).toBe('SESSION_NOT_FOUND');
            // And nothing was appended to B's conversation.
            expect(await store.listMessages(b.conversationId, 100)).toHaveLength(1);
        });

        it('verify_otp against B\'s session is SESSION_NOT_FOUND (no burning B\'s codes)', async () => {
            const store = new InMemorySessionStore();
            const b = await seed(store, 'b@example.com', 'bob secret');

            const { sink, dispatch } = connect(store, principal('a@example.com'));
            await dispatch({ action: 'verify_otp', requestId: 'vo', sessionId: b.sessionId, code: '123456' });

            expect((sink[0]!.error as Record<string, unknown>).code).toBe('SESSION_NOT_FOUND');
        });

        it('a principal with no email is denied too — emailless never means unrestricted', async () => {
            const store = new InMemorySessionStore();
            const b = await seed(store, 'b@example.com', 'bob secret');

            const { sink, dispatch } = connect(store, principal(undefined));
            await dispatch({ action: 'get_conversation_messages', requestId: 'gm', sessionId: b.sessionId });

            expect((sink[0]!.error as Record<string, unknown>).code).toBe('SESSION_NOT_FOUND');
        });

        it('the owner can still read their OWN session', async () => {
            const store = new InMemorySessionStore();
            const a = await seed(store, 'a@example.com', 'alice');

            const { sink, dispatch } = connect(store, principal('a@example.com'));
            await dispatch({ action: 'get_conversation_messages', requestId: 'gm', sessionId: a.sessionId });

            expect(sink[0]!.type).toBe('immediate_response');
            expect((sink[0]!.data as { messages: unknown[] }).messages).toHaveLength(1);
        });
    });

    describe('no existence oracle', () => {
        it("B's session and an id that never existed produce IDENTICAL responses", async () => {
            const store = new InMemorySessionStore();
            const b = await seed(store, 'b@example.com', 'bob secret');
            // Same shape/length as a real session id, so nothing but existence differs.
            const phantom = '00000000-0000-4000-8000-000000000000';

            for (const action of ['get_conversation_messages', 'get_session', 'send_message'] as const) {
                const real = connect(store, principal('a@example.com'));
                const fake = connect(store, principal('a@example.com'));
                await real.dispatch({ action, requestId: 'r', sessionId: b.sessionId, message: 'x' });
                await fake.dispatch({ action, requestId: 'r', sessionId: phantom, message: 'x' });

                // Byte-identical once the client-supplied id is normalized away: same code,
                // same message template, same frame shape. Any divergence is an oracle for
                // enumerating other users' session ids.
                const norm = (sink: Frame[], id: string) => JSON.stringify(sink).split(id).join('<ID>');
                expect(norm(real.sink, b.sessionId)).toBe(norm(fake.sink, phantom));
            }
        });

        it("resuming B's conversation behaves EXACTLY like resuming an unknown id — a fresh conversation", async () => {
            const store = new InMemorySessionStore();
            const b = await seed(store, 'b@example.com', 'bob secret');
            const phantom = '00000000-0000-4000-8000-000000000000';

            const real = connect(store, principal('a@example.com'));
            const fake = connect(store, principal('a@example.com'));
            await real.dispatch({ action: 'create_conversation_session', requestId: 'cs', agentId: 'agent-1', conversationId: b.conversationId });
            await fake.dispatch({ action: 'create_conversation_session', requestId: 'cs', agentId: 'agent-1', conversationId: phantom });

            const realData = real.sink[0]!.data as { sessionId: string; conversationId: string };
            const fakeData = fake.sink[0]!.data as { sessionId: string; conversationId: string };

            // Both mint a FRESH conversation rather than binding. Erroring on B's id while
            // silently minting for an unknown one would itself confirm B's id is real.
            expect(realData.conversationId).not.toBe(b.conversationId);
            expect(fakeData.conversationId).not.toBe(phantom);
            expect(real.sink[0]!.type).toBe(fake.sink[0]!.type);
            expect(real.sink[0]!.status).toBe(fake.sink[0]!.status);
            // And A's new session sees none of B's history.
            expect(await store.listMessages(realData.conversationId, 100)).toHaveLength(0);
        });
    });

    describe('the principal wins over the frame', () => {
        it("a client claiming someone else's userEmail does NOT get that user's scope", async () => {
            const store = new InMemorySessionStore();
            const b = await seed(store, 'b@example.com', 'bob secret');

            // A authenticates as themselves but lies in the create frame.
            const { sink, dispatch } = connect(store, principal('a@example.com'));
            await dispatch({
                action: 'create_conversation_session',
                requestId: 'cs',
                agentId: 'agent-1',
                userName: 'Bob',
                userEmail: 'b@example.com',
            });
            const spoofed = sink[0]!.data as { conversationId: string };
            await store.appendMessage(spoofed.conversationId, 'inbound', 'spoof attempt');

            // The spoofed session is owned by A (the principal), so it lands in A's list...
            const asA = connect(store, principal('a@example.com'));
            await asA.dispatch({ action: 'list_conversations', requestId: 'lc' });
            expect(conversationIds(asA.sink)).toContain(spoofed.conversationId);

            // ...and A still cannot see B's conversation.
            expect(conversationIds(asA.sink)).not.toContain(b.conversationId);

            // Nor does the spoofed session pollute B's list.
            const asB = connect(store, principal('b@example.com'));
            await asB.dispatch({ action: 'list_conversations', requestId: 'lc' });
            expect(conversationIds(asB.sink)).toEqual([b.conversationId]);
        });

        it('a resume cannot rewrite ownership of a conversation', async () => {
            const store = new InMemorySessionStore();
            const a = await seed(store, 'a@example.com', 'alice');

            // A legitimately resumes their own conversation while claiming B's email.
            const { sink, dispatch } = connect(store, principal('a@example.com'));
            await dispatch({
                action: 'create_conversation_session',
                requestId: 'cs',
                agentId: 'agent-1',
                conversationId: a.conversationId,
                userEmail: 'b@example.com',
            });
            expect((sink[0]!.data as { conversationId: string }).conversationId).toBe(a.conversationId);

            // B still can't see it.
            const asB = connect(store, principal('b@example.com'));
            await asB.dispatch({ action: 'list_conversations', requestId: 'lc' });
            expect(conversationIds(asB.sink)).toEqual([]);
        });
    });
});

File: go/server/user_scoping_test.go
package server

import (
	"context"
	"reflect"
	"sort"
	"testing"
)

// Adversarial coverage for th-8fe998: conversations are scoped to the connection's
// AUTHENTICATED principal. These tests are written from the attacker's side — user A holding
// a valid token trying to see, resume, or read user B's conversations — because the bug being
// fixed (list_conversations returning every user's conversations) passed every test written
// from the honest user's side.

// authedDispatcher builds a dispatcher for a connection authenticated as email, as an
// auth-ENABLED server would resolve it.
func authedDispatcher(store SessionStore, email string) *FrameDispatcher {
	access := AccessContext{Principal: Principal{Sub: email, Org: "acme", Email: email}, AuthEnabled: true}
	return NewFrameDispatcher(store, nil, access, "", nil, nil, nil, nil, nil, "", nil, nil, nil, nil)
}

// scopeFor is the scope an auth-enabled connection for email carries.
func scopeFor(email string) ConversationScope { return ConversationScope{Email: email} }

// seedConversation creates a conversation owned by ownerEmail with one inbound message, and
// returns the session it was created through.
func seedConversation(t *testing.T, store SessionStore, ownerEmail, text string) StoredSession {
	t.Helper()
	s, err := store.CreateSession(context.Background(), "agent", "U", ownerEmail, scopeFor(ownerEmail))
	if err != nil {
		t.Fatalf("create session for %s: %v", ownerEmail, err)
	}
	if _, err := store.AppendMessage(context.Background(), s.ConversationID, Inbound, text); err != nil {
		t.Fatalf("append for %s: %v", ownerEmail, err)
	}
	return s
}

// listConversationIDs drives list_conversations as email and returns the ids it returned.
func listConversationIDs(t *testing.T, store SessionStore, email string) []string {
	t.Helper()
	sink, events := capture()
	dispatchJSON(t, authedDispatcher(store, email), map[string]any{"action": "list_conversations", "requestId": "r1"}, sink)
	if len(*events) != 1 {
		t.Fatalf("want 1 event, got %+v", *events)
	}
	convs := (*events)[0]["data"].(map[string]any)["conversations"].([]map[string]any)
	ids := make([]string, 0, len(convs))
	for _, c := range convs {
		ids = append(ids, c["conversationId"].(string))
	}
	return ids
}

func TestListConversationsScopedToPrincipal(t *testing.T) {
	store := NewInMemorySessionStore()
	a := seedConversation(t, store, "a@example.com", "alice secret")
	b := seedConversation(t, store, "b@example.com", "bob secret")

	got := listConversationIDs(t, store, "a@example.com")
	if !reflect.DeepEqual(got, []string{a.ConversationID}) {
		t.Fatalf("A's list = %v, want exactly [%s] — B's conversation must never appear", got, a.ConversationID)
	}

	got = listConversationIDs(t, store, "b@example.com")
	if !reflect.DeepEqual(got, []string{b.ConversationID}) {
		t.Fatalf("B's list = %v, want exactly [%s]", got, b.ConversationID)
	}
}

// The original bug returned every user's conversations. With many other users present, a
// scoped list must still return the caller's own — not an empty page — which is what a filter
// applied AFTER the limit would produce.
func TestListConversationsFiltersBeforeLimit(t *testing.T) {
	store := NewInMemorySessionStore()
	// 60 conversations belonging to other people, more than defaultListLimit (50).
	for i := 0; i < 60; i++ {
		seedConversation(t, store, "other@example.com", "noise")
	}
	mine := seedConversation(t, store, "a@example.com", "mine")

	got := listConversationIDs(t, store, "a@example.com")
	if !reflect.DeepEqual(got, []string{mine.ConversationID}) {
		t.Fatalf("scoped list = %v, want exactly [%s]; a filter applied after the limit "+
			"would have returned an empty page here", got, mine.ConversationID)
	}
}

// getConversationMessagesEvent drives get_conversation_messages as email and returns the raw
// single event, so tests can compare whole payloads.
func getConversationMessagesEvent(t *testing.T, store SessionStore, email, sessionID string) map[string]any {
	t.Helper()
	sink, events := capture()
	dispatchJSON(t, authedDispatcher(store, email), map[string]any{
		"action": "get_conversation_messages", "requestId": "r1", "sessionId": sessionID,
	}, sink)
	if len(*events) != 1 {
		t.Fatalf("want 1 event, got %+v", *events)
	}
	return (*events)[0]
}

func TestGetConversationMessagesDeniesOtherUsersSession(t *testing.T) {
	store := NewInMemorySessionStore()
	b := seedConversation(t, store, "b@example.com", "bob secret")

	ev := getConversationMessagesEvent(t, store, "a@example.com", b.SessionID)
	if ev["type"] != "error" || errCode(ev) != "SESSION_NOT_FOUND" {
		t.Fatalf("A reading B's session got %+v, want a SESSION_NOT_FOUND error", ev)
	}
	// The leak the fix exists to close: not one byte of B's messages may reach A.
	if data, ok := ev["data"].(map[string]any); ok {
		if _, leaked := data["messages"]; leaked {
			t.Fatalf("denied read leaked messages: %+v", ev)
		}
	}
}

// THE EXISTENCE-ORACLE TEST. If "someone else's session" and "a session id that never
// existed" produce even slightly different responses, an attacker can enumerate valid session
// ids by diffing responses — so this asserts on the FULL payloads, not just the error code.
func TestNotYoursIsIndistinguishableFromNotFound(t *testing.T) {
	store := NewInMemorySessionStore()
	b := seedConversation(t, store, "b@example.com", "bob secret")

	notYours := comparable(getConversationMessagesEvent(t, store, "a@example.com", b.SessionID))
	neverExisted := comparable(getConversationMessagesEvent(t, store, "a@example.com", "00000000-0000-0000-0000-000000000000"))

	if !reflect.DeepEqual(notYours, neverExisted) {
		t.Fatalf("responses differ — this is an existence oracle for enumerating session ids.\n"+
			"  not-yours:      %+v\n  never-existed:  %+v", notYours, neverExisted)
	}

	// Same requirement on get_session, which reads the same store primitive.
	sinkA, evA := capture()
	dispatchJSON(t, authedDispatcher(store, "a@example.com"),
		map[string]any{"action": "get_session", "requestId": "r1", "sessionId": b.SessionID}, sinkA)
	sinkB, evB := capture()
	dispatchJSON(t, authedDispatcher(store, "a@example.com"),
		map[string]any{"action": "get_session", "requestId": "r1", "sessionId": "nope"}, sinkB)
	if a, b := comparable((*evA)[0]), comparable((*evB)[0]); !reflect.DeepEqual(a, b) {
		t.Fatalf("get_session is an existence oracle.\n  not-yours: %+v\n  unknown:   %+v", a, b)
	}
	if len(*evA) != len(*evB) {
		t.Fatalf("event COUNT differs (%d vs %d) — that alone is an oracle", len(*evA), len(*evB))
	}
}

// Resuming someone else's conversation must behave exactly like resuming an id that never
// existed: a brand-new, empty conversation. Anything else (an error, the real id echoed back)
// tells the attacker their guess was real.
func TestResumeOtherUsersConversationIsIndistinguishableFromUnknown(t *testing.T) {
	ctx := context.Background()
	store := NewInMemorySessionStore()
	b := seedConversation(t, store, "b@example.com", "bob secret")

	stolen, resumed, err := store.ResumeSession(ctx, "agent", "U", "a@example.com", scopeFor("a@example.com"), b.ConversationID)
	if err != nil {
		t.Fatalf("resume: %v", err)
	}
	if resumed {
		t.Fatal("A resumed B's conversation — cross-user resume must never bind")
	}
	if stolen.ConversationID == b.ConversationID {
		t.Fatalf("A's session bound to B's conversation %q", b.ConversationID)
	}
	if msgs, _ := store.ListMessages(ctx, stolen.ConversationID, 0); len(msgs) != 0 {
		t.Fatalf("A's fallback conversation carries history: %+v", msgs)
	}
	// B's conversation must be untouched — not re-homed onto A.
	if got := listConversationIDs(t, store, "b@example.com"); !reflect.DeepEqual(got, []string{b.ConversationID}) {
		t.Fatalf("B's conversation was disturbed by A's resume attempt: %v", got)
	}

	// And over the wire, the two cases are byte-identical apart from the freshly minted ids.
	steal := createSessionData(t, store, "a@example.com", b.ConversationID)
	unknown := createSessionData(t, store, "a@example.com", "does-not-exist")
	if steal["conversationId"] == b.ConversationID {
		t.Fatalf("create_conversation_session echoed B's conversation id back to A")
	}
	if !reflect.DeepEqual(keysOf(steal), keysOf(unknown)) {
		t.Fatalf("response shapes differ: %v vs %v", keysOf(steal), keysOf(unknown))
	}
}

// createSessionData drives create_conversation_session as email and returns its data payload.
func createSessionData(t *testing.T, store SessionStore, email, conversationID string) map[string]any {
	t.Helper()
	sink, events := capture()
	dispatchJSON(t, authedDispatcher(store, email), map[string]any{
		"action": "create_conversation_session", "requestId": "r1",
		"agentId": "agent", "conversationId": conversationID,
	}, sink)
	if len(*events) != 1 || (*events)[0]["type"] != "immediate_response" {
		t.Fatalf("want immediate_response, got %+v", *events)
	}
	return (*events)[0]["data"].(map[string]any)
}

// errCode pulls the error code out of an error event's descriptor.
func errCode(ev map[string]any) string {
	d, ok := ev["error"].(map[string]any)
	if !ok {
		return ""
	}
	code, _ := d["code"].(string)
	return code
}

// comparable strips the wall-clock timestamp every event carries, so two responses can be
// compared for the thing that actually matters: whether they are distinguishable in CONTENT.
// The timestamp varies between any two calls and is not an oracle.
func comparable(ev map[string]any) map[string]any {
	out := make(map[string]any, len(ev))
	for k, v := range ev {
		if k == "timestamp" {
			continue
		}
		out[k] = v
	}
	return out
}

func keysOf(m map[string]any) []string {
	out := make([]string, 0, len(m))
	for k := range m {
		out = append(out, k)
	}
	sort.Strings(out)
	return out
}

// THE SPOOFING TEST. userEmail is a client-supplied frame field. A caller who types someone
// else's address into it must NOT inherit that person's scope — the connection's principal
// decides ownership, and nothing the client sends can override it.
func TestClientSuppliedEmailCannotStealScope(t *testing.T) {
	store := NewInMemorySessionStore()
	victim := seedConversation(t, store, "victim@example.com", "victim secret")

	// Attacker authenticates as themselves but claims the victim's email in the frame.
	sink, events := capture()
	dispatchJSON(t, authedDispatcher(store, "attacker@example.com"), map[string]any{
		"action": "create_conversation_session", "requestId": "r1",
		"agentId": "agent", "userName": "Victim", "userEmail": "victim@example.com",
	}, sink)
	if len(*events) != 1 {
		t.Fatalf("want 1 event, got %+v", *events)
	}

	// The spoofed email must not have widened what the attacker can list.
	got := listConversationIDs(t, store, "attacker@example.com")
	for _, id := range got {
		if id == victim.ConversationID {
			t.Fatalf("client-supplied userEmail granted the victim's scope: %v", got)
		}
	}

	// Nor may it have stamped the victim as owner of the attacker's new session.
	sid := (*events)[0]["data"].(map[string]any)["sessionId"].(string)
	s, _ := store.GetSession(context.Background(), sid)
	if s.OwnerEmail != "attacker@example.com" {
		t.Fatalf("OwnerEmail = %q, want the authenticated principal %q — ownership must come "+
			"from the token, not the frame", s.OwnerEmail, "attacker@example.com")
	}
	// The client-supplied address survives only as the OTP delivery contact.
	if s.ContactEmail != "victim@example.com" {
		t.Fatalf("ContactEmail = %q, want the frame value preserved for OTP delivery", s.ContactEmail)
	}
}

// Auth enabled but the token carries no email: no OWNED conversation is visible. Such a
// principal owns nothing and can therefore reach nothing that anyone else owns — the leak a
// well-meaning "if email == \"\" { return everything }" fallback would open.
//
// Owner-LESS conversations are a separate population and DO stay reachable (th-909995): this
// principal's own sessions are stamped with no owner, so denying them means denying it the
// product entirely. TestEmaillessPrincipalCanConverseOnAuthEnabledServer covers that side.
func TestAuthEnabledWithoutEmailSeesNoOwnedConversations(t *testing.T) {
	store := NewInMemorySessionStore()
	a := seedConversation(t, store, "a@example.com", "alice secret")
	unowned, _ := store.CreateSession(context.Background(), "agent", "U", "", ConversationScope{Unscoped: true})
	_, _ = store.AppendMessage(context.Background(), unowned.ConversationID, Inbound, "auth-disabled era")

	emailless := AccessContext{Principal: Principal{Sub: "svc", Org: "acme"}, AuthEnabled: true}
	d := NewFrameDispatcher(store, nil, emailless, "", nil, nil, nil, nil, nil, "", nil, nil, nil, nil)

	sink, events := capture()
	dispatchJSON(t, d, map[string]any{"action": "list_conversations", "requestId": "r1"}, sink)
	convs := (*events)[0]["data"].(map[string]any)["conversations"].([]map[string]any)
	for _, c := range convs {
		if c["conversationId"] == a.ConversationID {
			t.Fatalf("emailless principal saw A's OWNED conversation: %+v", convs)
		}
	}

	// A's session itself stays hidden; the owner-less one is reachable.
	for _, tc := range []struct {
		name    string
		id      string
		wantErr bool
	}{
		{"owned by A", a.SessionID, true},
		{"owner-less", unowned.SessionID, false},
	} {
		sink, events := capture()
		dispatchJSON(t, d, map[string]any{
			"action": "get_conversation_messages", "requestId": "r1", "sessionId": tc.id,
		}, sink)
		denied := errCode((*events)[0]) == "SESSION_NOT_FOUND"
		if denied != tc.wantErr {
			t.Fatalf("emailless principal on session %s: denied=%v, want denied=%v (%+v)",
				tc.name, denied, tc.wantErr, (*events)[0])
		}
	}
}

// Auth disabled (no verifier configured) is the ONE unscoped path and must keep working
// exactly as before — this is the local/dev single-tenant case.
func TestAuthDisabledStaysUnscoped(t *testing.T) {
	ctx := context.Background()
	store := NewInMemorySessionStore()
	unscoped := ConversationScope{Unscoped: true}
	one, _ := store.CreateSession(ctx, "agent", "U", "u1@example.com", unscoped)
	_, _ = store.AppendMessage(ctx, one.ConversationID, Inbound, "first")
	two, _ := store.CreateSession(ctx, "agent", "U", "u2@example.com", unscoped)
	_, _ = store.AppendMessage(ctx, two.ConversationID, Inbound, "second")

	// bareDispatcher carries the zero AccessContext — AuthEnabled false — the no-auth server.
	sink, events := capture()
	dispatchJSON(t, bareDispatcher(store), map[string]any{"action": "list_conversations", "requestId": "r1"}, sink)
	convs := (*events)[0]["data"].(map[string]any)["conversations"].([]map[string]any)
	if len(convs) != 2 {
		t.Fatalf("auth-disabled list returned %d conversations, want both (2)", len(convs))
	}

	// Resume across sessions still binds, and reads still succeed.
	if _, resumed, _ := store.ResumeSession(ctx, "agent", "U", "u2@example.com", unscoped, one.ConversationID); !resumed {
		t.Fatal("auth-disabled resume stopped binding an existing conversation")
	}
	ev := map[string]any{"action": "get_conversation_messages", "requestId": "r1", "sessionId": two.SessionID}
	sink2, events2 := capture()
	dispatchJSON(t, bareDispatcher(store), ev, sink2)
	if (*events2)[0]["type"] != "immediate_response" {
		t.Fatalf("auth-disabled read denied: %+v", (*events2)[0])
	}
}

func TestConversationScopeAllows(t *testing.T) {
	tests := []struct {
		name  string
		scope ConversationScope
		owner string
		want  bool
	}{
		{"unscoped sees owned", ConversationScope{Unscoped: true}, "a@example.com", true},
		{"unscoped sees owner-less", ConversationScope{Unscoped: true}, "", true},
		{"match", scopeFor("a@example.com"), "a@example.com", true},
		{"match ignores email casing", scopeFor("Alice@Example.COM"), "alice@example.com", true},
		{"mismatch", scopeFor("a@example.com"), "b@example.com", false},
		// Owner-less conversations have nobody to enforce on behalf of, so they stay
		// reachable — otherwise anonymous and emailless principals are locked out of the
		// sessions they themselves created. th-909995 (Option B).
		{"scoped sees owner-less", scopeFor("a@example.com"), "", true},
		{"zero value denies OWNED conversations", ConversationScope{}, "a@example.com", false},
		{"zero value sees owner-less", ConversationScope{}, "", true},
	}
	for _, tc := range tests {
		t.Run(tc.name, func(t *testing.T) {
			if got := tc.scope.Allows(tc.owner); got != tc.want {
				t.Fatalf("Allows(%q) = %v, want %v", tc.owner, got, tc.want)
			}
		})
	}
}

// The verifier must distinguish "no auth configured" from "auth configured, token rejected".
// Collapsing the two hands a forged token the unscoped view of everything.
func TestVerifierScopeDerivation(t *testing.T) {
	if s := (PermissiveVerifier{}).Resolve("").ConversationScope(); !s.Unscoped {
		t.Fatal("no-auth verifier must yield the unscoped scope")
	}

	v := NewLocalTokenVerifier("shhh")
	for _, tok := range []string{"", "garbage", "a.b.c", "not.a.jwt.at.all"} {
		access := v.Resolve(tok)
		if !access.AuthEnabled {
			t.Fatalf("rejected token %q lost AuthEnabled — it would be treated as no-auth", tok)
		}
		s := access.ConversationScope()
		if s.Unscoped || s.Email != "" {
			t.Fatalf("rejected token %q yielded scope %+v, want fail-closed zero value", tok, s)
		}
		if s.Allows("a@example.com") {
			t.Fatalf("rejected token %q can still see another user's OWNED conversation", tok)
		}
	}
}

func TestClaimsToAccessCarriesEmail(t *testing.T) {
	access, err := claimsToAccess([]byte(`{"sub":"u1","org":"acme","email":" a@example.com "}`))
	if err != nil {
		t.Fatalf("claims: %v", err)
	}
	if access.Principal.Email != "a@example.com" {
		t.Fatalf("Email = %q, want the trimmed claim", access.Principal.Email)
	}
	if !access.AuthEnabled {
		t.Fatal("a verified token must be AuthEnabled")
	}
	if got := access.ConversationScope(); got.Unscoped || got.Email != "a@example.com" {
		t.Fatalf("scope = %+v, want scoped to a@example.com", got)
	}

	// A token with no email claim must NOT be defaulted into some placeholder identity.
	noEmail, err := claimsToAccess([]byte(`{"sub":"u1","org":"acme"}`))
	if err != nil {
		t.Fatalf("claims: %v", err)
	}
	if noEmail.Principal.Email != "" {
		t.Fatalf("Email = %q, want empty (no default identity)", noEmail.Principal.Email)
	}
	if s := noEmail.ConversationScope(); s.Unscoped {
		t.Fatal("emailless token widened to unscoped — this is the fail-open the fix forbids")
	}
}
